philosophy/FORMS.md

```diff
@@ -0,0 +1,928 @@
+error like so:
+
+```js
+{
+  values: { name: 'Party Parrot', slug: 'party-parrot' },
+  touched: { name: true, slug: true },
+  errors: { slug: { duplicate: true } }
+}
+```
+
+We can then use this information to render the error. This solves the problem of
+mapping async errors back to the form.
+
+Of course it's possible to also build async validation while the user fills out
+the form, but this check is the basis of that. It needs to be there in any case,
+as the slug could get duplicated between the time we validate the users input and
+the time the user actually submits the form.
+
+Something else which could be improved is to validate the slug is using valid
+characters only. We can use our old `validate` function for that. The errors
+can be attached to `errors.slug`.
+
+### Single inputs to determine multiple values (e.g. `MoneyInput`)
+
+Sometimes we want to build input elements which deal with multiple values at once.
+One such element is our [`MoneyInput`](https://uikit.commercetools.com/?selectedKind=Inputs&selectedStory=MoneyInput). It allows to set a currency and an amount by using a dropdown for the currency and an input for the amount.
+
+The value produced by `MoneyInput` is represented as `{ currencyCode: 'EUR', amount: '2.00' }`
+
+> Sidenote: As the `MoneyInput` uses monetary formatting, we are using a text-input. This allows us to format values properly (e.g. by adding the trailing ".00") without browsers messing with the input. It also allows us to use decimal separators, e.g. for "1,000.00".
+
+We can construct our Formik value like this:
+
+```js
+{
+  id: 'party-parrot',
+  version: 1,
+  name: 'Party Parrot',
+  price: { currencyCode: 'EUR', amount: '2.00' }
+}
+```
+
+How do we now build a single input which receives `{ currencyCode: 'EUR', amount: '2.00' }` as its value and is able to change both using Formik's `handleChange`? We want to be able to render `<MoneyInput onChange={handleChange} onBlur={handleBlur} />` to make it as straightforward as using `<input type="text" />`.
+
+There are two approaches:
+
+- We can ensure `MoneyInput` always calls `onChange` with a full value like `{ currencyCode: 'EUR', amount: '2.00' }`
+- We can adapt the input names to be `price.currencyCode` or `price.amount` respectively. Formik will read that name and set the nested value only.
+
+If we were to go with the first approach, we'd need to fake `event.target.name` in `onBlur` as well, otherwise formik would set `{ touched: { price: true } }` instead of `{ touched: { price: { amount: true, currencyCode: true } } }`. That is bad, as Formik will set ``{ touched: { price: { amount: true, currencyCode: true } } }`once the form is submitted - so we end up with a`touched`state which changes shape! There is a CodeSandbox which illustrates this problem. Try changing the price and keep an eye on the`touched`state. Then submit the form, and you'll see a different`touched` state.
```

Comment:
```suggestion
If we were to go with the first approach, we'd need to fake `event.target.name` in `onBlur` as well, otherwise formik would set `{ touched: { price: true } }` instead of `{ touched: { price: { amount: true, currencyCode: true } } }`. That is bad, as Formik will set `{ touched: { price: { amount: true, currencyCode: true } } }`once the form is submitted - so we end up with a `touched` state which changes shape! There is a CodeSandbox which illustrates this problem. Try changing the price and keep an eye on the `touched` state. Then submit the form, and you'll see a different `touched` state.
```

Reply:
(the suggestion above fixes some bad formatting with backticks)